src/Text/Pandoc/Readers/Markdown.hs

```diff
@@ -1741,13 +1742,41 @@ source = do
 linkTitle :: MarkdownParser String
 linkTitle = quotedTitle '"' <|> quotedTitle '\''
 
-link :: MarkdownParser (F Inlines)
-link = try $ do
+wikilinkInfo :: MarkdownParser (F Inlines, String)
+wikilinkInfo = do
+  string "[["
+  (_, raw) <- withRaw $ charsInBalancedBrackets 2
+  guard $ not $ null raw
+  let deunder c = if c == '_' then ' ' else c
```

Comment:
A lot of transformations are performed on the parsed string.  The alternative, which would probably be easier to read, would be to shift this logic into the parsing stage.  It requires a slightly different way to think, but Parsec is a joy to work with.  If you can, give it a try.